pkg/kubelet/config/apiserver.go

```diff
@@ -27,7 +27,14 @@ import (
 )
 
 // NewSourceApiserver creates a config source that watches and pulls from the apiserver.
-func NewSourceApiserver(client *client.Client, hostname string, updates chan<- interface{}) {
+func NewSourceApiserver(clientInterface client.Interface, hostname string, updates chan<- interface{}) {
+	// TODO: client.Interface doesn't have required methods to deal with ListWatch; it also doesn't provide
+	// method to get client version.
+	client, ok := clientInterface.(*client.Client)
```

Comment:
Why not just build a ListWatch directly?  Pass your list func and your watch func. 

I think we should kill NewListWatchFromClient altogether.

Sent from my iPhone

> On Mar 5, 2015, at 5:00 PM, Deyuan Deng <notifications@github.com> wrote:
> 
> In pkg/kubelet/config/apiserver.go:
> 
> > @@ -27,7 +27,14 @@ import (
> >  )
> >  
> >  // NewSourceApiserver creates a config source that watches and pulls from the apiserver.
> > -func NewSourceApiserver(client *client.Client, hostname string, updates chan<- interface{}) {
> > +func NewSourceApiserver(clientInterface client.Interface, hostname string, updates chan<- interface{}) {
> > +	// TODO: client.Interface doesn't have required methods to deal with ListWatch; it also doesn't provide
> > +	// method to get client version.
> > +	client, ok := clientInterface.(*client.Client)
> I think this is a larger problem than an unfortunate TODO.
> 
> Our client interface doesn't support 1. listing via fields; 2. get client version. Even if it does, like we said earlier, we can list/watch for a particular resource. Which means if we remove the type assertion here, we'll add it in cache.NewListWatchFromClient.
> 
> —
> Reply to this email directly or view it on GitHub.
>